ext/backtracie_native_extension/public/backtracie.h

```diff
@@ -0,0 +1,187 @@
+// (at your option) any later version.
+//
+// backtracie is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with backtracie.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef BACKTRACIE_PUBLIC_H
+#define BACKTRACIE_PUBLIC_H
+
+#include <ruby.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#define BACKTRACIE_ABI_VERSION ((uint32_t)0x1)
+
+// clang-format off
+#if defined(_WIN32) || defined(_WIN64)
+# ifdef BACKTRACIE_EXPORTS
+#   define BACKTRACIE_API __declspec(dllexport)
+# else
+#   define BACKTRACIE_API __declspec(dllimport)
+# endif
+#else
+# define BACKTRACIE_API __attribute__((visibility("default")))
+#endif
+// clang-format on
+
+typedef struct {
+  // The first element of this struct is a bitfield, comprised of uint32_t's.
+  // Ruby requires that VALUE be at least 32-bit but not nescessarily 64-bit;
+  // using uint32_t for the bitfield ensures we have a minimum of pointless
+  // padding in this struct no matter the system.
+
+  // This needs to be set by the caller so that Backtracie knows the size of
+  // this struct, and can remain ABI compatible if a new backtracie gem is
+  // compiled and linked against an app expecting an older version.
+  uint32_t abi_version : 8;
+  // Whether or not the frame is "valid"; if zero, the entire rest of the struct
+  // is completley meaningless.
+  uint32_t is_valid : 1;
+  // 1 -> ruby frame / 0 -> cfunc frame
+  uint32_t is_ruby_frame : 1;
+  // 1 means self really is the self object, 0 means it's the class of the self.
+  uint32_t self_is_real_self : 1;
+  // The iseq & cme; one, both, or neither might actually be available; if
+  // they're not, they will be Qnil. These need to be GC marked if you wish to
+  // keep the location alive.
+  VALUE iseq;
+  VALUE callable_method_entry;
+  // This is either self, or rb_class_of(self), depending on a few thing.
+  // It will be the actual self object for the frame if
+  //   - self is the toplevel binding,
+  //   - self is rb_mRubyVMFrozenCore,
+  //   - self is a class or module
+  // Otherwise, it will be rb_class_of(self).
+  // This is done so that, should the caller decide to retain the raw_location
+  // instance for a while, GC mark its VALUEs, and stringify the backtrace
+  // later, we both:
+  //   - retain enough information to actually produce a good name for the
+  //   method,
+  //   - but also don't hold onto random objects that would otherwise be GC'd
+  //   jsut
+  //     because they happened to be in a backtrace.
+  // This needs to be marked if you wish to keep the location alive
+  VALUE self_or_self_class;
+  // Raw PC pointer; not much use to callers, but saved so we can calculate the
+  // lineno later on.
+  const void *pc;
```

Comment:
Minor: Not sure if relevant, but Ruby usually stores this as a `VALUE *`.

Reply:
I did think about storing this as a VALUE *, but we do nothing with this other than pointer arithmetic with it, and I think I wanted it to be opaque to callers (I don't _think_ there's anything sensible a user of the library could do with *pc?). Keeping it in here as a void pointer hopefully stops people from getting any ideas about trying to use it for some nefarious purpose ;)

Reply:
Sounds good :+1: 

Ahaha somehow with what we're doing with backtracie I don't think we can judge other people for doing weird things any more :)